src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -394,12 +431,16 @@ namespace detail
                     {
                         /// If error is not retriable -- false is returned and exception is set.
                         /// Otherwise the error is thrown and retries continue.
-                        bool initialized = initialize();
-                        if (!initialized)
+                        auto error = initialize();
```

Comment:
In case of initialization in constructor and getting `SKIP_NOT_FOUND_URL` we will try to initialize here again (because `impl` is still nullptr) and we will get `SKIP_NOT_FOUND_URL` here again. Maybe remember the error from initialization in constructor?